pkg/metastore/metastore.go

```diff
@@ -18,13 +24,28 @@ func (m *Table) Update(db, table, serde string, transformLocation func(location
 	m.TableName = table
 	m.Sd.SerdeInfo.Name = serde
 	var err error
-	if m.Sd.Location != "" {
+	if m.Sd.Location != "" && !(isSparkSQLTable && strings.HasSuffix(m.Sd.Location, sparkSQLWorkaroundSuffix)) {
 		m.Sd.Location, err = transformLocation(m.Sd.Location)
 	}
+	if err != nil {
+		return err
+	}
+
+	if isSparkSQLTable {
+		// Table was created by Spark SQL, we should change the internal stored Spark SQL location
+		if l, ok := m.Sd.SerdeInfo.Parameters[sparkSQLProviderLocationKey]; ok {
+			m.Sd.SerdeInfo.Parameters[sparkSQLProviderLocationKey], err = transformLocation(l)
+		}
+	}
```

Comment:
I think if we build a `func UpdateStorageDescriptor(sd *StorageDescriptor) (*StorageDescriptor, error)` and pass it to the table and partition, it will move this logic into one place where we select the way we like to address the change and how it will be applied.